Test: Allow reporting results as JUnit

reneme opened the pull request:
Based on the refactoring in #3027 this adds an `XmlReporter` that emits [JUnit5](https://github.com/junit-team/junit5/blob/main/platform-tests/src/test/resources/jenkins-junit.xsd)-ish XML test results. The results are written to disk into a directory specified via `./botan-test --test-results-dir=...` as `Botan-3.0.0-alpha0-tests-<current timestamp>.xml`. "ne can provide additional properties via `--report-properties=key:value,schluessel:wert`.

Eventually, we're planning to visualize the reports into pull requests using external tools, like:
* https://www.check-run-reporter.com/
* https://github.com/mikepenz/action-junit-report

### Example Output

` ./botan-test --test-results-dir=test-results --report-properties=foo:bar block`

```xml
<?xml version="1.0" encoding="UTF-8"?>
<testsuites tests="27" failures="0" time="0.268">
<properties>
<property name="CPU flags" value="sse2 ssse3 sse41 sse42 avx2 avx512f avx512dq avx512bw avx512_icelake rdtsc bmi1 bmi2 adx aes_ni clmul rdrand rdseed intel_sha avx512_aes avx512_clmul" />
<property name="current test run" value="1" />
<property name="drbg_seed" value="172A3B0A6E75E755" />
<property name="foo" value="bar" />
<property name="timestamp" value="2022-11-23T14:52:07+0100" />
<property name="total test runs" value="1" />
</properties>
<testsuite name="block" tests="27" failures="0" timestamp="2022-11-23T14:52:07+0100" time="0.256">
<testcase name="AES-128" assertions="17235" timestamp="2022-11-23T14:52:07+0100" time="0.021" file="src/tests/test_block.cpp" line="199" />
<testcase name="AES-192" assertions="20250" timestamp="2022-11-23T14:52:07+0100" time="0.023" file="src/tests/test_block.cpp" line="199" />
<testcase name="AES-256" assertions="23130" timestamp="2022-11-23T14:52:07+0100" time="0.028" file="src/tests/test_block.cpp" line="199" />
<testcase name="ARIA-128" assertions="30" timestamp="2022-11-23T14:52:07+0100" time="0.000" file="src/tests/test_block.cpp" line="199" />
<testcase name="ARIA-192" assertions="30" timestamp="2022-11-23T14:52:07+0100" time="0.000" file="src/tests/test_block.cpp" line="199" />
<testcase name="ARIA-256" assertions="30" timestamp="2022-11-23T14:52:07+0100" time="0.000" file="src/tests/test_block.cpp" line="199" />
<testcase name="Blowfish" assertions="930" timestamp="2022-11-23T14:52:07+0100" time="0.007" file="src/tests/test_block.cpp" line="199" />
<testcase name="CAST-128" assertions="720" timestamp="2022-11-23T14:52:07+0100" time="0.001" file="src/tests/test_block.cpp" line="199" />
<testcase name="Camellia-128" assertions="90" timestamp="2022-11-23T14:52:07+0100" time="0.000" file="src/tests/test_block.cpp" line="199" />
<testcase name="Camellia-192" assertions="45" timestamp="2022-11-23T14:52:07+0100" time="0.000" file="src/tests/test_block.cpp" line="199" />
<testcase name="Camellia-256" assertions="75" timestamp="2022-11-23T14:52:07+0100" time="0.000" file="src/tests/test_block.cpp" line="199" />
<testcase name="Cascade(Serpent,AES-256)" assertions="30" timestamp="2022-11-23T14:52:07+0100" time="0.000" file="src/tests/test_block.cpp" line="199" />
<testcase name="Cascade(Serpent,CAST-128)" assertions="15" timestamp="2022-11-23T14:52:07+0100" time="0.000" file="src/tests/test_block.cpp" line="199" />
<testcase name="Cascade(Serpent,Twofish)" assertions="45" timestamp="2022-11-23T14:52:07+0100" time="0.000" file="src/tests/test_block.cpp" line="199" />
<testcase name="DES" assertions="4815" timestamp="2022-11-23T14:52:07+0100" time="0.004" file="src/tests/test_block.cpp" line="199" />
<testcase name="GOST-28147-89(R3411_94_TestParam)" assertions="270" timestamp="2022-11-23T14:52:07+0100" time="0.000" file="src/tests/test_block.cpp" line="199" />
<testcase name="GOST-28147-89(R3411_CryptoPro)" assertions="150" timestamp="2022-11-23T14:52:07+0100" time="0.000" file="src/tests/test_block.cpp" line="199" />
<testcase name="IDEA" assertions="16260" timestamp="2022-11-23T14:52:07+0100" time="0.023" file="src/tests/test_block.cpp" line="199" />
<testcase name="Lion(SHA-160,RC4,64)" assertions="15" timestamp="2022-11-23T14:52:07+0100" time="0.000" file="src/tests/test_block.cpp" line="199" />
<testcase name="Noekeon" assertions="15450" timestamp="2022-11-23T14:52:07+0100" time="0.014" file="src/tests/test_block.cpp" line="199" />
<testcase name="SEED" assertions="60" timestamp="2022-11-23T14:52:08+0100" time="0.000" file="src/tests/test_block.cpp" line="199" />
<testcase name="SHACAL2" assertions="61260" timestamp="2022-11-23T14:52:08+0100" time="0.064" file="src/tests/test_block.cpp" line="199" />
<testcase name="SM4" assertions="45" timestamp="2022-11-23T14:52:08+0100" time="0.000" file="src/tests/test_block.cpp" line="199" />
<testcase name="Serpent" assertions="47070" timestamp="2022-11-23T14:52:08+0100" time="0.049" file="src/tests/test_block.cpp" line="199" />
<testcase name="Threefish-512" assertions="150" timestamp="2022-11-23T14:52:08+0100" time="0.000" file="src/tests/test_block.cpp" line="199" />
<testcase name="TripleDES" assertions="840" timestamp="2022-11-23T14:52:07+0100" time="0.001" file="src/tests/test_block.cpp" line="199" />
<testcase name="Twofish" assertions="16545" timestamp="2022-11-23T14:52:08+0100" time="0.021" file="src/tests/test_block.cpp" line="199" />
</testsuite>
</testsuites>
```

reneme commented:
I tried [mikepenz/action-junit-report](https://github.com/mikepenz/action-junit-report) and with a few more tweaks this can [annotate failing test cases right into the code](https://github.com/reneme/botan/pull/1/files#diff-452a4a7c46333522991d7f9d22c0b81e58709f553645709d26270792f818ec0e):

![Screenshot 2022-08-05 121419](https://user-images.githubusercontent.com/1562139/183056412-afb4fccb-f937-4470-8357-01f587f7af0e.png)

Obviously, this works only if the failure information contains code locations: I.e. we would need to use some macro-tricks to make this happen. (And touch all the test code to incorporate the macros).

But even without that, it generates test results in the GitHub actions summary view:
![Screenshot 2022-08-05 121751](https://user-images.githubusercontent.com/1562139/183056984-5c7a96a9-3c49-470f-b607-fee6654cc0aa.png)

All of this comes at the expense of using two 3rd party GitHub Actions:
* [mikepenz/action-junit-report](https://github.com/mikepenz/action-junit-report) and
* [dawidd6/action-download-artifact](https://github.com/dawidd6/action-download-artifact)
Plus [an additional workflow YAML](https://github.com/mikepenz/action-junit-report#pr-run-permissions) to orchestrate it all and keep the workflow access permissions in control.

reneme commented:
Now for check-run-reporter.com: This [produces a somewhat more comprehensive report](https://github.com/reneme/botan/pull/2) as a single summary page behind a link in the "Checks"-Box of a Pull Request:

![Screenshot 2022-08-05 172553](https://user-images.githubusercontent.com/1562139/183109788-56add653-ea30-44f0-8247-360b933c3d5a.png)

---

![Screenshot 2022-08-05 171710](https://user-images.githubusercontent.com/1562139/183108016-ad66e87e-9eb2-4cb6-aeea-a1b2e9b550eb.png)

Though, if the same test fails for more than one build configuration it seems to overwrite previous fails, which might or might not be problematic.

On the plus side, it's much easier to set up and can work without a custom 3rd-party action. I.e. one can manually `curl -X POST` the generated JUnit results to the service.

It has a free-tier for open source projects that is capped to 10.000 monthly reports and 1GB of transfer. That may or may not be enough.

Bottom line: neither of these two services is really convincing unfortunately.
@randombit What do you think? Do you see this bringing enough added value?

Frankly: I'm somewhat feeling the urge to write a simple action that is tailored to our needs... Particularly considering other to-be-integrated reports from BoGo, Python-based tests and so on.

reneme commented on src/tests/runner/test_reporter.h:
```diff
@@ -12,6 +12,58 @@
+      std::optional<std::chrono::nanoseconds> elapsed_time() const;
+
+      const std::string& name() const { return m_name; }
+      const std::vector<TestSummary> results() const { return m_results; }
```
That is because it was supposed to be a reference. Good catch!

reneme commented on src/tests/runner/test_reporter.cpp:
```diff
@@ -11,6 +11,69 @@
+   {
+   if(!a.has_value() || !b.has_value())
+      {
+      return std::nullopt;
```
That's a fairly philosophical question: Is `std::nullopt` equal to `0` (for the integer case)?
I'd argue: no, its different. To me, that's one of the main selling points of `std::optional<>`: A distinct and unambiguous way to declare something as "not existing". And if something doesn't exist, we shouldn't do math with It, I would say.

lieser replied:
As this is essentially only a helper I think the question we should ask is how `Testsuite::elapsed_time()` should behave.

If one of the `TestSummary` in `m_results` is a `std::nullopt`, should `Testsuite::elapsed_time()`:
- return a  `std::nullopt` (current behavior)
- ignore it and still return the accumulated time of the other tests.

So I just wanted to make sure that it is intended that a single `std::nullopt` in the results means `Testsuite::elapsed_time()` will not return a value for elapsed time.

reneme replied:
I looked into that again. As stated, the test runner makes sure that `elapsed_time` is set for all tests. I considered to check that with a `BOTAN_ASSERT` but refrained from it; after all, test results with missing time measurements are still more valuable than a crashing test executable.

For the sum: in the case that not all time measurements are available (which should be considered a bug), I find it reasonable that the aggregate time is also not defined. After all 3s + 2s + **undefined** s isn't really 5s but an undefined time greater than 5s, no? A missing aggregate value should be considered a bug that is worth tracking down.

Bottom line: My vote is on leaving it "as is".

reneme commented on src/tests/runner/test_reporter.h:
```diff
@@ -12,6 +12,58 @@
+class TestSummary final
+   {
+   public:
+      TestSummary(const Test::Result& result);
```
True, that bothers me as well. I introduced this additional type to have an explicit and light-weight way of summarizing test results. I believe there was a more technical reason that would have needed further refactoring. OTOH I can't remember, though. :(

reneme replied:
One reason: The `Test::Result` is what all the tests interact with, and I wanted to keep this "summarizing" interface away from the tests.

reneme commented on src/tests/runner/test_xml_reporter.h:
```diff
@@ -0,0 +1,57 @@
+   private:
+      std::string get_unique_output_filename() const;
+
+      void render_preamble(std::ostream& out) const;
```
I wanted to make this as flexible as possible. As it turns out, just using the member would probably have worked, too. 

One minor detail: `m_outfile` is a `std::optional<>` to reflect that the output file might not be open. By taking a plain `std::ostream&` as a parameter, the top-most `render()` method can deal with the state handling of the `std::optional<>` and thus leaves this concern out of the downstream render methods.

reneme commented on src/tests/runner/test_xml_reporter.cpp:
```diff
@@ -0,0 +1,286 @@
+      {
+      std::stringstream ss;
+      ss << "Failed to open '" << file << "' for writing JUnit report.";
+      throw Botan::Invalid_Argument(ss.str());
```
True. `Botan::System_Error` fits better.

> This exception is thrown in the event of an error related to interacting with the operating system.

reneme commented:
Rebased onto https://github.com/randombit/botan/pull/3058 and added the code locations to the XML report. Hence. we should first review/merge the code locations and then rebase this again onto master.

randombit commented:
I know there are some unresolved questions here wrt reporting actions and how to improve the location reporting, but is there any reason this PR couldn't merge now?

reneme commented:
Indeed there's no show stopper. The JUnit reports just won't be used as is. But we certainly can come back to that later.

reneme commented:
I incoporated the changes and fixes I needed here: https://github.com/Rohde-Schwarz/botan/pull/10 . Mainly, the addition of `./botan-test --report-properties=key:value,other_key:other_value` to set additional properties in `ci_build.py` that can later be used when generating human-readable reports from the JUnit files.

From my side, this is ready to go in now.

lieser commented on src/tests/runner/test_xml_reporter.cpp:
```diff
@@ -0,0 +1,308 @@
+
+void XmlReporter::next_run()
+   {
+   if(m_outfile.has_value() && m_outfile->good())
```
Why only explicitly close the file if it is good?

reneme replied:
Only "good" files may be eligible for graceful handling. 🤡

Changed it to `m_outfile->is_open()`

lieser replied:
You did change a different place inside the `XmlReporter::~XmlReporter()`. Question is also if in the destructor we maybe could drop it completely. Destructor of `std::basic_ofstream` should close the file automatically.
If we do not check if the close is successful we don't get anything from an explicit close.

reneme replied:
🤦‍♂️

Yes, certainly: We won't need this destructor. Same in the place you commented in the first place. `std::optional<>::reset()` will destruct the wrapped `std::ofstream` if need be and that will 'do the right thing'™.

reneme commented on src/scripts/ci_build.py:
```diff
@@ -110,6 +110,19 @@ def determine_flags(target, target_os, target_cpu, target_cc, cc_bin,
 
+    # generate JUnit test report
+    if test_results_dir:
+        if not os.path.isdir(test_results_dir):
+            raise Exception("Test results directory does not exist")
```
I did want to create it here at first. But then I noticed that `determine_flags()` didn't seem to have any side effects on the file system and wanted to preserve that. I went for the check to safeguard future refactorings. No strong feelings about this one, though.

lieser replied:
ok, also fine with leaving as is

reneme commented on src/scripts/ci_build.py:
```diff
@@ -110,6 +110,19 @@ def determine_flags(target, target_os, target_cpu, target_cc, cc_bin,
+        if not os.path.isdir(test_results_dir):
+            raise Exception("Test results directory does not exist")
+
+        def escape(some_string):
+            return some_string.replace(':', '').replace(',', '')
```
I agree.

reneme commented on src/tests/runner/test_runner.cpp:
```diff
@@ -89,6 +91,14 @@ bool Test_Runner::run(const Test_Options& opts)
+#if defined(BOTAN_TARGET_OS_HAS_FILESYSTEM)
+      m_reporters.emplace_back(std::make_unique<XmlReporter>(opts, opts.xml_results_dir()));
+#else
+      output() << "Generating test report files is not supported on this platform\n";
```
That's debatable indeed.

That way we don't need to add special treatment for build configurations without file system support in the CI scripts (i.e. omit the `--test-results-dir=` parameter). We simply don't get a report from such a configuration.

While that behavior might be somewhat confusing, the final result of "not having a report for this build configuration" is the same. When running into this interactively, I would hope the user would notice the warning, though.

... as I said: debatable. I don't feel very strongly either way.

lieser replied:
Ok with leaving as is if it makes the matrix builds easier.
